### PR TITLE
docs: use swagger for api documentation

### PR DESCRIPTION
**Pull request by domenkozar**
#### TODO

- [x] [servant-swagger-ui](https://github.com/phadej/servant-swagger-ui/blob/master/example/Main.hs)
- [x] [servant-auth-swagger](https://github.com/plow-technologies/servant-auth/pull/36)

2017-03-14-144356_1062x926_scrot

cc @jkarni @expipiplus1

### CODE REVIEW

**Comment by domenkozar** [comment]
It fails due to nixpkgs not having bullt all dependenices, need to wait a bit more :)

**Comment by domenkozar** [comment]
Here's the rendered API:

2017-03-14-144356_1062x926_scrot

**Comment by expipiplus1** [comment]
@domenkozar Cool! looks good!

**Comment by jkarni** [comment]
Looks cool indeed!

By the way, you can add annotations to the doc as a whole or individual endpoints. There's an example [here](https://hackage.haskell.org/package/servant-swagger-1.1.2.1/docs/Servant-Swagger.html#g:3)

**Comment by domenkozar** [comment]
TODO:

- [ ] swagger ui
- [ ] servant-auth-swagger

**Comment by domenkozar** [comment]
Almost there, these errors are (still) very confusing to me:

```
src/Hercules/Lib.hs:70:14: error:
    • Ambiguous type variables ‘m0’, ‘dir0’, ‘api0’ arising from a use of ‘enter’
      prevents the constraint ‘(servant-0.9.1.1:Servant.Utils.Enter.Enter
                                  (m0 (ServerT
                                         api0
                                         (transformers-0.5.2.0:Control.Monad.Trans.Except.ExceptT
                                            ServantErr IO)
                                       :<|> (transformers-0.5.2.0:Control.Monad.Trans.Except.ExceptT
                                               ServantErr IO (SwaggerUiHtml dir0 api0)
                                             :<|> (transformers-0.5.2.0:Control.Monad.Trans.Except.ExceptT
                                                     ServantErr IO (SwaggerUiHtml dir0 api0)
                                                   :<|> Application))))
                                  (App
                                   :~> transformers-0.5.2.0:Control.Monad.Trans.Except.ExceptT
                                         ServantErr IO)
                                  (transformers-0.5.2.0:Control.Monad.Trans.Except.ExceptT
                                     ServantErr IO Swagger
                                   :<|> (transformers-0.5.2.0:Control.Monad.Trans.Except.ExceptT
                                           ServantErr
                                           IO
                                           (SwaggerUiHtml
                                              "" ("swagger.json" :> Get '[JSON] Swagger))
                                         :<|> (transformers-0.5.2.0:Control.Monad.Trans.Except.ExceptT
                                                 ServantErr
                                                 IO
                                                 (SwaggerUiHtml
                                                    "" ("swagger.json" :> Get '[JSON] Swagger))
                                               :<|> Application))))’ from being solved.
      Relevant bindings include
        api :: forall (dir :: ghc-prim-0.5.0.0:GHC.Types.Symbol) b b1.
               ((App [Text]
                 :<|> (App [Project]
                       :<|> ((Text -> App (Maybe Project))
                             :<|> App [ProjectWithJobsets])))
                :<|> (AuthResult UserId -> App Text))
               :<|> ((App Text.Blaze.Html.Html
                      :<|> ((AuthenticatorName
                             -> Maybe Hercules.OAuth.Types.AuthClientState
                             -> Maybe Hercules.OAuth.Types.FrontendURL
                             -> App b1)
                            :<|> ((AuthenticatorName
                                   -> Maybe AuthCode
                                   -> Maybe Hercules.OAuth.Types.AuthError
                                   -> Maybe Hercules.OAuth.Types.AuthStatePacked
                                   -> App b)
                                  :<|> ((Maybe Text -> App Text.Blaze.Html.Html)
                                        :<|> (AuthResult UserId -> App Text.Blaze.Html.Html)
))))
                     :<|> m0 (ServerT
                                api0
                                (transformers-0.5.2.0:Control.Monad.Trans.Except.ExceptT
                                   ServantErr IO)
                              :<|> (transformers-0.5.2.0:Control.Monad.Trans.Except.ExceptT
                                      ServantErr IO (SwaggerUiHtml dir api0)
                                    :<|> (transformers-0.5.2.0:Control.Monad.Trans.Except.Ex
ceptT
                                            ServantErr IO (SwaggerUiHtml dir api0)
                                          :<|> Application))))
          (bound at src/Hercules/Lib.hs:71:9)
      Probable fix: use a type annotation to specify what ‘m0’, ‘dir0’, ‘api0’ should be.
      These potential instance exist:
        instance (servant-0.9.1.1:Servant.Utils.Enter.Enter typ1 arg1 ret1,
                  servant-0.9.1.1:Servant.Utils.Enter.Enter typ2 arg2 ret2,
                  arg1 ~ arg2) =>
                 servant-0.9.1.1:Servant.Utils.Enter.Enter
                   (typ1 :<|> typ2) arg1 (ret1 :<|> ret2)
          -- Defined in ‘servant-0.9.1.1:Servant.Utils.Enter’
    • In the expression: enter (Nat (runApp env)) api
      In an equation for ‘server’:
          server env
            = enter (Nat (runApp env)) api
            where
                api
                  = queryApi
                    :<|> pages :<|> (return $ swaggerSchemaUIServer swaggerDoc)
                pages
                  = welcomePage
                    :<|>
                      (mandatory1 .: loginPage)
                      :<|>
                        (mandatory1 .∵ authCallback)
                        :<|> loggedInPage :<|> (join . withAuthenticated userInfoPage)
                queryApi = unprotected :<|> protected
                unprotected
                  = getProjectNames
                    :<|> getProjects :<|> getProject :<|> getProjectsWithJobsets
                ....

src/Hercules/Lib.hs:73:30: error:
    • Couldn't match type ‘ServerT api0 Servant.Handler’
                     with ‘transformers-0.5.2.0:Control.Monad.Trans.Except.ExceptT
                             ServantErr IO Swagger’
        arising from a use of ‘swaggerSchemaUIServer’
      The type variable ‘api0’ is ambiguous
    • In the second argument of ‘($)’, namely
        ‘swaggerSchemaUIServer swaggerDoc’
      In the second argument of ‘(:<|>)’, namely
        ‘(return $ swaggerSchemaUIServer swaggerDoc)’
      In the second argument of ‘(:<|>)’, namely
        ‘pages :<|> (return $ swaggerSchemaUIServer swaggerDoc)’
    • Relevant bindings include
        api :: ((App [Text]
                 :<|> (App [Project]
                       :<|> ((Text -> App (Maybe Project))
                             :<|> App [ProjectWithJobsets])))
                :<|> (AuthResult UserId -> App Text))
               :<|> ((App Text.Blaze.Html.Html
                      :<|> ((AuthenticatorName
                             -> Maybe Hercules.OAuth.Types.AuthClientState
                             -> Maybe Hercules.OAuth.Types.FrontendURL
                             -> App b1)
                            :<|> ((AuthenticatorName
                                   -> Maybe AuthCode
                                   -> Maybe Hercules.OAuth.Types.AuthError
                                   -> Maybe Hercules.OAuth.Types.AuthStatePacked
                                   -> App b)
                                  :<|> ((Maybe Text -> App Text.Blaze.Html.Html)
                                        :<|> (AuthResult UserId -> App Text.Blaze.Html.Html)
))))
                     :<|> m0 (Server (SwaggerSchemaUI' dir api0)))
          (bound at src/Hercules/Lib.hs:71:9)

```

**Comment by domenkozar** [comment]
@expipiplus1 ready for review.

File: backend/src/Hercules/API.hs [hunk 1]
```diff
@@ -47,5 +48,7 @@ type Pages = "login" :> Get '[HTML] Html
                          :> Get '[HTML] Html
         :<|> "repos" :> Auth '[JWT] UserId :> Get '[HTML] Html
 
-type API = QueryAPI
+type API = (QueryAPI
       :<|> Pages
+      :<|> Get '[HTML] Html) -- https://github.com/haskell-servant/servant/issues/117
```
Comment (expipiplus1) [inline backend/src/Hercules/API.hs]: Might want to explain why this is referenced here and what this html does (redirects always).

Comment (domenkozar) [inline backend/src/Hercules/API.hs]: Done :)

**Pull request merged**